crates/ev-prover/src/server/mod.rs

```diff
@@ -31,138 +32,258 @@ use crate::prover::service::ProverService;
-    #[allow(unused_mut)]
-    let (tx_range, rx_range) = mpsc::channel::<MessageProofRequest>(256);
-    let message_sync = MessageProofSync::shared();
-
     #[cfg(not(feature = "combined"))]
     {
+        let message_sync = MessageProofSync::shared();
         let batch_size = config_clone.batch_size;
         let concurrency = config_clone.concurrency;
         let queue_capacity = config_clone.queue_capacity;
-        let (tx_block, rx_block) = mpsc::channel(256);
-        tokio::spawn({
-            let block_prover = BlockExecProver::new(
-                AppContext::new(config_clone, trusted_state)?,
-                tx_block,
-                storage.clone(),
-                queue_capacity,
-                concurrency,
-            );
-            async move {
-                if let Err(e) = block_prover.run().await {
-                    error!("Block prover task failed: {e:?}");
-                }
+        let (tx_range, rx_range) = mpsc::channel::<MessageProofRequest>(256);
+        let (tx_block, rx_block) = mpsc::channel::<BlockProofCommitted>(256);
+        let block_prover = BlockExecProver::new(
+            AppContext::new(config_clone.clone(), trusted_state)?,
+            tx_block,
+            storage.clone(),
+            queue_capacity,
+            concurrency,
+        );
+        let block_range_prover = BlockRangeExecProver::new()?;
+        let message_prover = prepare_message_prover(reth_rpc_url, reth_ws_url, storage.clone())?;
+        let server = Server::new(
+            Arc::new(message_prover),
+            Arc::new(block_prover),
+            Arc::new(block_range_prover),
+        );
+        let mut block_handle = server.start_block_prover().await?;
+        let mut message_handle = server.start_message_prover(rx_range, ism_client, message_sync).await?;
+        let mut block_range_handle = server
+            .start_block_range_prover(client, storage.clone(), rx_block, tx_range, batch_size)
+            .await?;
+        tokio::select! {
+            r = &mut block_handle => {
+                error!("block prover stopped: {:?}", r);
+                message_handle.abort();
+                block_range_handle.abort();
             }
-        });
-
-        let prover = Arc::new(BlockRangeExecProver::new()?);
-        let service =
-            BlockRangeExecService::new(client, prover, storage.clone(), rx_block, tx_range, batch_size, 16).await?;
-
-        tokio::spawn(async move {
-            if let Err(e) = service.run().await {
-                error!("Block prover task failed: {e:?}");
+            r = &mut message_handle => {
+                error!("message prover stopped: {:?}", r);
+                block_handle.abort();
+                block_range_handle.abort();
             }
-        });
+            r = &mut block_range_handle => {
+                error!("block range prover stopped: {:?}", r);
+                block_handle.abort();
+                message_handle.abort();
```

Comment:
so you mean we abort, and re-initialize? where is the re-initialization happening?